Implement simd_as for pointers

calebzulawski opened the pull request:
Expands `simd_as` (and `simd_cast`) to handle pointer-to-pointer, pointer-to-integer, and integer-to-pointer conversions.

cc @programmerjake @thomcc

rust-highfive commented:
r? @michaelwoerister

(rust-highfive has picked a reviewer for you, use r? to override)

programmerjake commented on compiler/rustc_codegen_llvm/src/intrinsic.rs:
```diff
@@ -1722,40 +1722,33 @@ unsupported {} from `{}` with element `{}` of size `{}` to `{}`"#,
 
-        let (in_style, in_width) = match in_elem.kind() {
+        let determine_style = |kind: &ty::TyKind<'_>| match kind {
             // vectors of pointer-sized integers should've been
```
this comment is out-of-date -- it should probably just be removed

calebzulawski replied:
Good point, removed

programmerjake reviewed (approved):
lgtm

workingjubilee commented:
Hm. This behavior should probably be a separate intrinsic rather than be behind `simd_as` or `simd_cast`, because the intention is that the pointer to integer cast will also become a separate intrinsic for Rust scalar code (to reflect its current separation in the existing Rust pointer API of `addr` and `with_addr`, et cetera... those will eventually become backed by intrinsics, instead of using `as`).

programmerjake commented:
> Hm. Unless I am misunderstanding which part of the interface you are trying to model here, this behavior should probably be a separate intrinsic rather than be behind `simd_as` or `simd_cast`, because the intention is that the pointer to integer cast will also become a separate intrinsic for Rust scalar code (to reflect its current separation in the existing Rust pointer API of `addr` and `with_addr`, et cetera... those will eventually become backed by intrinsics, instead of using `as`).

imho we need a simd equivalent to `as` as is currently used by scalar code for int <-> ptr casts, imho the strict_provenance things would be their own intrinsics, but wouldn't replace `as` since that's used for things like  `expose_addr` and `from_exposed_addr`.

programmerjake commented:
> Hm. Unless I am misunderstanding which part of the interface you are trying to model here, this behavior should probably be a separate intrinsic rather than be behind `simd_as` or `simd_cast`, because the intention is that the pointer to integer cast will also become a separate intrinsic for Rust scalar code (to reflect its current separation in the existing Rust pointer API of `addr` and `with_addr`, et cetera... those will eventually become backed by intrinsics, instead of using `as`).

imho we need a simd equivalent to `as` as is currently used by scalar code for int <-> ptr casts, imho the strict_provenance things would be their own intrinsics, but wouldn't replace `as` since that's used for things like  `expose_addr` and `from_exposed_addr`.

calebzulawski commented:
Yeah, there is no intention to make integer `as` pointer etc invalid.  Since there currently is no new intrinsic, it's not really within the scope of portable SIMD to determine what that new intrinsic is.  This change is necessary at a minimum to remove some transmutes from `std::simd` which are actually UB.

workingjubilee commented:
It's not clear to me we necessarily want the default cast to use the semantics of `as`, however, because `as` will actually be the cast that invalidates more optimizations.

cc @RalfJung

RalfJung commented:
> there is no intention to make integer as pointer etc invalid

Not sure what you mean by "invalid". At least some people intent to basically deprecate `as` because it has so many problems. For things like `i8 as i32`, the intended replacement is `From`/`Into`; for `ptr as usize`, the replacement is `expose_addr`, for `usize as ptr`, it is `from_exposed_addr`.

Given all the problems with `as` in scalar code, I am not sure if it is the best idea to copy it for SIMD. And given all the subtleties around pointer/integer casts, I think a more explicit API such as `expose_addr`/`from_exposed_addr` is much preferable over the extremely implicit `as`.

calebzulawski commented:
I suppose by "intention" I meant that it's obviously a commonly used feature and requires an edition change, and future compilers will still need it to support old editions.  Maybe in a (fairly far away) future edition of rust there will be no `as`, but I don't think it makes sense to wait until then for a perfect implementation when `std::simd` has UB today, which can be fixed.

Without `as`, how do I cast between pointer types?

RalfJung commented:
> Without as, how do I cast between pointer types?

https://doc.rust-lang.org/nightly/std/primitive.pointer.html#method.cast
This is safer since it cannot accidentally also change mutability.

We still need a method to explicitly change mutability. I didn't say we are there yet.

And indeed `as` casts will not be removed, but they might become discouraged and maybe even deprecated.

---

Anyway, I am mostly concerned about casts between pointers and integers. Those are very different from the other casts, and that's why we now have dedicated methods with very detailed documentation for [ptr2int casts](https://doc.rust-lang.org/nightly/std/primitive.pointer.html#method.expose_addr) and [int2ptr casts](https://doc.rust-lang.org/nightly/std/ptr/fn.from_exposed_addr.html).

calebzulawski commented:
It's worth noting that I intend on implementing (and have already partially) the strict provenance API for std::simd--I just can't add some of these functions without a cast intrinsic.

RalfJung commented:
Right, this is just the intrinsic, nothing user-facing.

However, I don't think it is a good idea to lump together `expose_addr` and `from_exposed_addr` with the other 'as'-casts. We are already having trouble with that in MIR; some optimizations that are valid for other casts are not valid for these two (or at least, not valid for `expose_addr`). That's why they are now a separate [`CastKind`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.CastKind.html).

So IMO it would also be better to make their SIMD versions separate intrinsics.

RalfJung commented:
Or should the semantics of these be that of `addr` and `ptr::invalid`? Those are not available as `as`-casts in scalar Rust. They have to be written via `transmute` (or by calling the unstable Strict Provenance methods).

calebzulawski commented:
I really just need the exposing casts, I've already implemented the others via transmute.  Are there equivalent scalar pointer cast intrinsics I can reference? I implemented it with the `as` name because the comments in `std` seemed to indicate that there are no such intrinsics yet.

RalfJung commented:
They are not exposed as intrinsics, they are (for historic reasons) exposed via `as`. However, they *are* a separate CastKind, and several MIR passes (and Miri) need to treat them different from all the other `as` casts.

IMO they should be separate `Rvalue` variants in MIR, but we haven't done that refactoring yet. No reason for SIMD to repeat the same mistakes, though.

calebzulawski commented:
This is well above my head.  Everything in portable SIMD is done via intrinsics, so I'm not really sure what the alternative is

RalfJung commented:
I think what you should do is to have a dedicated intrinsic for "SIMD expose_addr", and another dedicated intrinsic for "SIMD from_exposed_addr", and have `simd_as` *not* support casting between pointers and integers.

That very roughly corresponds to how these operations are/should also be represented differently in MIR.

michaelwoerister commented:
r? rust-lang/libs (or is this even something for the lang team to discuss?)

calebzulawski commented:
I've changed this PR to add 3 new pointer intrinsics, rather than overload `simd_as`.

programmerjake reviewed (approved):
lgtm

RalfJung commented on compiler/rustc_codegen_llvm/src/intrinsic.rs:
```diff
@@ -1704,6 +1704,85 @@ unsupported {} from `{}` with element `{}` of size `{}` to `{}`"#,
+        );
+
+        match in_elem.kind() {
+            ty::RawPtr(_) => {}
```
Is it worth checking that this is a ptr to a sized type, or more genreally not a wide ptr?

calebzulawski replied:
I suppose that should really be rejected by `repr(simd)` (I'm not sure if it is)

calebzulawski replied:
Also, how are wide casts handled by the backend, I wonder? Like `*const [T] as *const [U]`?  If we ever had vectors of fat pointers, I suppose that would still be something valid.

RalfJung replied:
The backend code for casts has tons of special cases to cover all the possibilities.

calebzulawski replied:
Just confirmed, `repr(simd)` does already reject wide pointers: https://github.com/rust-lang/rust/blob/ffa77332c630b43cb92f6256b8e404198820fa03/compiler/rustc_middle/src/ty/layout.rs#L909-L919

RalfJung replied:
Okay so you don't think hat is worth double-checking here? (I usually prefer not to rely on checks that are done in other crates.)

calebzulawski replied:
I'm a little hesitant since checking if it's sized here seems to require the parameter environment etc.  Either way, if `repr(simd)` is changed to allow wide pointers, `simd_arith_offset` needs to be updated as well.

calebzulawski replied:
I read through some of the docs and this was much more of a non-issue than I thought, so I added in a check to ensure the pointed-to types are sized when casting.

programmerjake commented on compiler/rustc_codegen_llvm/src/intrinsic.rs:
```diff
@@ -1704,6 +1704,93 @@ unsupported {} from `{}` with element `{}` of size `{}` to `{}`"#,
+
+        match in_elem.kind() {
+            ty::RawPtr(p) => require!(
+                p.ty.is_sized(bx.tcx.at(span), ty::ParamEnv::reveal_all()),
```
imho this should check for [thin](https://github.com/rust-lang/rust/issues/81513) pointers rather than pointers to sized types, we should be able to have vectors of pointers to extern types. Afaict you need to call [`ptr_metadata_ty`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Ty.html#method.ptr_metadata_ty) and handle the results. Maybe ask on Zulip if there's a handy `is_thin_ptr` function or something...

calebzulawski replied:
I believe extern types are represented by `Opaque`, and this line is hinting that they must be sized, anyway:
https://github.com/rust-lang/rust/blob/c07a8b4e09f356c7468b69c50cac7fc5b5000b8a/compiler/rustc_middle/src/ty/sty.rs#L2123

programmerjake replied:
extern types are `ty::Foreign` and unsized, with metadata `()`:
https://github.com/rust-lang/rust/blob/5b8cf49c51833ee5d27ae2e8e179337dbb9f14d7/compiler/rustc_type_ir/src/sty.rs#L52-L53

imho you should not check specifically for `ty::Foreign` and instead use `ptr_metadata_ty`:
```rust
// probably right, should be double-checked though
let (metadata_ty, need_sized_check) = pointee_ty.ptr_metadata_ty(...);
if need_sized_check {
    // idk how to do this
    require_sized_trait(pointee_ty);
}
// now you can use metadata_ty
if !is_unit_type(metadata_ty) {
    // error
}
```
something like that...though maybe a better route is to get the in-memory layout of the pointer type and check that it's just a single pointer.

calebzulawski replied:
I'm having a bit of trouble figuring out what "normalize" in this context means.  I only see one use of this function anywhere in rustc...

programmerjake replied:
I asked on Zulip: https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/help.20checking.20for.20thin.20pointers.20in.20codegen/near/295534523

calebzulawski replied:
I believe it should be good, now

thomcc commented:
@michaelwoerister Hmm, why did you request review from libs? This is all compiler code.

RalfJung commented:
intrinsics are lang territory though. *Exposing* intrinsics is libs territory, but that's not what this PR does.

thomcc commented:
I can r+ this on behalf of libs, but not the lang/compiler parts.

RalfJung commented:
None of these simd intrinsics are exposed on stable, so I don't think lang has been involved much so far. I expect that will be all discussed together when (some subset of) `std::simd` moves towards stabilization.

RalfJung commented:
r? compiler

wesleywiser commented:
r? compiler

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 3f2ce0624dfed866f758521b225e26c00b3250d8 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 3f2ce0624dfed866f758521b225e26c00b3250d8 -->
<!-- homu: {"type":"Approved","sha":"3f2ce0624dfed866f758521b225e26c00b3250d8","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

calebzulawski commented:
Thanks all!
